controller/k8s/api.go

```diff
@@ -49,46 +51,47 @@ func NewLister(k8sClient kubernetes.Interface) *Lister {
 	replicationControllerInformer := sharedInformers.Core().V1().ReplicationControllers()
 	serviceInformer := sharedInformers.Core().V1().Services()
 
-	lister := &Lister{
-		NS:     namespaceInformer.Lister(),
-		Deploy: deployInformer.Lister(),
-		RS:     replicaSetInformer.Lister(),
-		Pod:    podInformer.Lister(),
-		RC:     replicationControllerInformer.Lister(),
-		Svc:    serviceInformer.Lister(),
+	api := &API{
+		NS:     namespaceInformer,
+		Deploy: deployInformer,
+		RS:     replicaSetInformer,
+		Pod:    podInformer,
+		RC:     replicationControllerInformer,
+		Svc:    serviceInformer,
 
 		nsSynced:     namespaceInformer.Informer().HasSynced,
 		deploySynced: deployInformer.Informer().HasSynced,
 		rsSynced:     replicaSetInformer.Informer().HasSynced,
 		podSynced:    podInformer.Informer().HasSynced,
 		rcSynced:     replicationControllerInformer.Informer().HasSynced,
 		svcSynced:    serviceInformer.Informer().HasSynced,
-	}
 
-	// this must be called after the Lister() methods
-	sharedInformers.Start(nil)
+		sharedInformers: sharedInformers,
+	}
 
-	return lister
+	return api
 }
 
 // Sync waits for all informers to be synced.
 // For servers, call this asynchronously.
 // For testing, call this synchronously.
-func (l *Lister) Sync() error {
+func (api *API) Sync() error {
+	api.sharedInformers.Start(nil)
```

Comment:
curious about moving `sharedInformers.Start()` from `NewAPI()` to `Sync()`. since `Sync()` is called asynchronously in our servers, i'm wondering if there could be a timing issue where our server processes could start trying to interact with the `k8s.API` object prior to `Sync()` (and thus `sharedInformers.Start()`) getting called.

Reply:
Good question -- what you describe is exactly the intent of this change. We can't call `sharedInformers.Start()` until all indexers and event handlers have been added. So for instance in #1073, the destination server adds an indexer by calling:

```
	k8sAPI.Pod.Informer().AddIndexers(cache.Indexers{"ip": indexPodByIp})
```

That call to `AddIndexers` has to happen before calling `sharedInformers.Start()`. This configuration wouldn't work if `sharedInformers.Start()` happened when the k8sAPI object was created.